sdk/src/main/java/io/radar/sdk/RadarBatteryManager.kt

```diff
@@ -0,0 +1,163 @@
+package io.radar.sdk
+
+import android.annotation.SuppressLint
+import android.app.usage.UsageStatsManager
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.os.BatteryManager
+import android.os.Build
+import android.os.PowerManager
+
+/**
+ * Handles battery-related information within the Radar SDK
+ */
+internal class RadarBatteryManager(
+    private val context: Context
+) {
+
+    private val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager?
+    private val locationUnaffected = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+        PowerManager.LOCATION_MODE_NO_CHANGE
+    } else {
+        0
+    }
+    private val usageStatsManager = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+        context.getSystemService(Context.USAGE_STATS_SERVICE) as UsageStatsManager?
+    } else {
+        null
+    }
+
+    /**
+     * Get the battery percentage of the device
+     *
+     * @see <a href="https://developer.android.com/training/monitoring-device-state/battery-monitoring">
+     *     Android Battery Monitoring</a>
+     */
+    fun getBatteryState(): BatteryState {
+        val batteryStatus: Intent? = IntentFilter(Intent.ACTION_BATTERY_CHANGED).let { filter ->
+            context.registerReceiver(null, filter)
+        }
+        val status: Int = batteryStatus?.getIntExtra(BatteryManager.EXTRA_STATUS, -1) ?: -1
+        val isCharging: Boolean = status == BatteryManager.BATTERY_STATUS_CHARGING
+                || status == BatteryManager.BATTERY_STATUS_FULL
+
+        val batteryPct: Float? = batteryStatus?.let { intent ->
+            val level: Int = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
+            val scale: Int = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
+            level * 100 / scale.toFloat()
+        }
+        return BatteryState(isCharging, batteryPct ?: 0F, getPerformanceState())
+    }
+
+    @SuppressLint("NewApi")
+    fun getAppStandbyBucket(): Int? = usageStatsManager?.appStandbyBucket
+
+    /**
+     * Get the performance state. This uses Android's [PowerManager] to gather information
+     * about the current optimization settings that could affect the SDK's performance, and returns
+     * a simplified status which identifies whether or not battery optimizations may be affecting
+     * the SDK's responsiveness.
+     */
+    private fun getPerformanceState(): PerformanceState {
+        val powerSaveMode = isPowerSaveMode()
+        if (powerSaveMode != null) {
```

Comment:
This decision tree is a nice way to see how Android makes use of some of these modes in practice.  One thing I'm wondering about is - how closely will various manufactures or OS versions adhere to this logic?  My hunch is that it might vary a little in important ways.  With that in mind, and to maximize observability, I think it could be really useful to, when we log the `PerformanceState` value, to also log the underlying flags that are determining it.  Looks like there might be 4-5 flags we could pass up.  We could learn alot from them I think.

Reply:
done